File: hash_tool.py
from pathlib import Path
import hashlib
from check_file_permission import check_file_permission

def hash_file(file_path, config, algorithm="sha256"):
    if not check_file_permission(file_path, "read", config):
        raise PermissionError(f"Read access denied for {file_path}")
    file_path = Path(file_path).resolve()
    if not file_path.is_file():
        raise ValueError(f"File not found: {file_path}")
    hasher = hashlib.new(algorithm)
    with open(file_path, "rb") as f:
        hasher.update(f.read())
    return hasher.hexdigest()

File: ts-tools/hash_tool.ts

import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
import { checkFilePermission } from './check_file_permission';

export function hashFile(filePath: string, config: any, algorithm: string = "sha256"): Promise<string> {
    return new Promise((resolve, reject) => {
        checkFilePermission(filePath, "read", config)
            .then(() => {
                const resolvedPath = path.resolve(filePath);
                const hash = crypto.createHash(algorithm);
                const stream = fs.createReadStream(resolvedPath);
                stream.on('data', (data) => {
                    hash.update(data);
                });
                stream.on('end', () => {
                    resolve(hash.digest('hex'));
                });
                stream.on('error', (err) => {
                    reject(err);
                });
            })
            .catch(reject);
    });
}
